lib/irb/color.rb

```diff
@@ -170,33 +165,37 @@ def without_circular_ref(obj, seen:, &block)
       end
 
       def scan(code, allow_last_error:)
-        pos = [1, 0]
-
         verbose, $VERBOSE = $VERBOSE, nil
         RubyLex.compile_with_errors_suppressed(code) do |inner_code, line_no|
           lexer = Ripper::Lexer.new(inner_code, '(ripper)', line_no)
+          byte_pos = 0
+          line_positions = [0]
+          inner_code.lines.each do |line|
+            line_positions << line_positions.last + line.bytesize
+          end
+
+          on_scan = proc do |elem|
```

Comment:
Do you think it's more understandable if we rewrite the block like this?

```rb
  on_scan = proc do |elem|
    start_pos = line_positions[elem.pos[0] - 1] + elem.pos[1]

    # yield uncolorable code
    if byte_pos < start_pos
      yield(nil, inner_code.byteslice(byte_pos...start_pos), nil)
    end

    if byte_pos <= start_pos
      str = elem.tok
      yield(elem.event, str, elem.state)
      byte_pos = start_pos + str.bytesize
    end
  end
```

Reasons for the changes:

1. It's easier to understand if we always put `byte_pos` and `start_pos` at the same position in comparisons.
2. We can move local variable assignment closer to where it's going to be used.
3. In complicated logic like this, I feel traditional if block is easier to understand.

Reply:
The code you suggested looks more readable. thanks